### ISSUE TITLE
Incompleteness issue in the configuration /arith: {2,5,6,7,8}

### ISSUE THREAD

**Issue by muchang** [dafny-lang/dafny#5005]
### Dafny version

4.4.0

### Code to produce this issue

```dafny
method main (){
    var tmp_0 := false;
    var tmp_2 := 2;
    var tmp_1 := 1 <= (- tmp_2);
    var i := 0; 
    while (tmp_1 && i < 10) {
        tmp_0 := true;
        i := i + 1;
    }
    assert (! tmp_0);
}
```

### Command to run and resulting output

```code
$ ./Dafny test.dfy /arith:2

test.dfy(10,11): Error: assertion might not hold

Dafny program verifier finished with 0 verified, 1 error
```

### What happened?

Dafny should verify this program because the variable `tmp_1` should be `false` then the variable `tmp_0` should be `false` at the end of the program. However, Dafny cannot verify this program when using /arith: {2,5,6,7,8}. Replacing `tmp_2` in `var tmp_1 := 1 <= (- tmp_2)` makes this incompleteness bug disappear. 

It seems to have happened since Dafny 3.13.0.   

### What type of operating system are you experiencing the problem on?

Linux, WSL

**Comment by keyboardDrummer** [comment]
If `/arith: {2,5,6,7,8}` turns off the built-in SMT theory for `-`, then wouldn't you expect `tmp_1` to have an unknown value, and Dafny not to be able to prove the assertion?

**Comment by atomb** [comment]
> If `/arith: {2,5,6,7,8}` turns off the built-in SMT theory for `-`, then wouldn't you expect using this to cause `tmp_1` to have an unknown value, and Dafny not to be able to prove the assertion?

Yes, I think that's correct. Not being able to prove this is the intentional semantics of those arithmetic modes. The reason for that is that you may have separately proved some lemmas about arithmetic operations and want to use those in place of the arithmetic reasoning that is built in to Z3, so these modes let you turn off that reasoning. In most cases, that's probably not what you want, though.

As to why you see it sometimes working: if you replace `tmp2` with `2` in the declaration for `tmp_1`, I suspect there's some constant folding going on that makes the solver no longer see this as a subtraction operation but rather a comparison with a negative constant.

**Comment by muchang** [comment]
I see, then it makes sense. Thanks!
